README.md
# advanced-fetch

[![GitHub license](https://img.shields.io/github/license/sigalor/advanced-fetch)](https://github.com/sigalor/advanced-fetch/blob/master/LICENSE) [![npm](https://img.shields.io/npm/v/advanced-fetch)](https://www.npmjs.com/package/advanced-fetch) [![Unit tests workflow status](https://github.com/sigalor/advanced-fetch/actions/workflows/tests.yaml/badge.svg)](https://github.com/sigalor/advanced-fetch/actions/workflows/tests.yaml)

A version of node-fetch with more convenient functionality.

## Installation

```
npm install advanced-fetch
```

## Usage

```javascript
const { Fetch } = require("advanced-fetch");

(async () => {
  const fetch = new Fetch();
  await fetch.get("http://example.com");
  // subsequent requests to example.com now send the received
  // cookies again, and the cookies are updated automatically
})();
```

## Constructor

The `Fetch` constructor expects an optional object, with the following possible attributes:

- `cookiesFilename`: If given, the [tough-cookie](https://www.npmjs.com/package/tough-cookie) cookie jar is serialized into JSON after each request and that JSON is written into the specified file.
- `encoding`: If given, the received responses are assumed to be in that text encoding and converted automatically to UTF-8 via [iconv-lite](https://www.npmjs.com/package/iconv-lite) before they are returned to you.

## Request parameters

Supports `.get`, `.post`, `.put` and `.delete`. After the URL, all of these functions also expect an optional object with [node-fetch](https://www.npmjs.com/package/node-fetch) request parameters.

Additionally, you can specify the following parameters as normal JavaScript objects, which are serialized respectively automatically:

- `query` for GET query parameters, internally serialized using [querystring](https://nodejs.org/api/querystring.html)
- `form` for the body of a `Content-Type: application/x-www-form-urlencoded` request, internally serialized using [URLSearchParams](https://nodejs.org/api/url.html#url_class_urlsearchparams)
- `formData` for the body of a `Content-Type: multipart/form-data` request, internally serialized using [form-data](https://www.npmjs.com/package/form-data)
- `json` for the body of a `Content-Type: application/json` request, internally serialized using `JSON.stringify`

Finally, you can specify `returnBuffer: true` in the parameters object in order to get a Node.js `Buffer` as the result of a request instead of a string.

## License

MIT


CHANGELOG.md
# 3.1.2 (2023-02-01)

- fix: handling of query parameter separator

# 3.1.1 (2023-01-10)

- also export all types and interfaces for TypeScript

# 3.1.0 (2023-01-10)

- do not use any default redirect mode anymore, i.e. use the default behavior of node-fetch in case nothing else is mentioned
- `AdvancedFetchResponse.urls` is not optional anymore

# 3.0.0 (2023-01-10)

- upgrade to node-fetch@3 via [node-fetch-commonjs](https://www.npmjs.com/package/node-fetch-commonjs)
- introduce new redirect mode called `followWithCookies` which is the default and where the redirects respect cookies
- change `returnBuffer: true` to `returnType: "buffer"`, with the following possible values for `returnType`: string (default), buffer, json (actual default when response has `Content-Type: application/json`)

# 2.2.1 (2023-01-10)

- rebuild library

# 2.2.0 (2023-01-10)

- add AdvancedFetchResponse.urls with list of redirected URLs

# 2.1.0 (2023-01-10)

- update NPM dependencies

# 2.0.0 (2022-09-02)

- fix array form parameters and encoding handling
- update NPM dependencies
- add `CHANGELOG.md` file
- add badges to readme
- change constructor usage from `await new Fetch(...)` to just `new Fetch()`
- add basic unit tests
- rewrite entire library in TypeScript

# 1.1.0 (2021-08-24)

- add `commonFetchParams` and `ignoreInvalidHttps` constructor parameters
- add `getCookie` and `requestWithFullResponse` utility functions
- update NPM dependencies

# 1.0.0 (2020-11-01)

- initial release


test/server.ts
import express from 'express';
import cookieParser from 'cookie-parser';
import bodyParser from 'body-parser';
import http from 'http';

const app = express();
app.use(cookieParser());
app.use(bodyParser.json());

let server: http.Server | undefined = undefined;

app.get('/', (req, res) => {
  res.send('Hello World!');
});

app.get('/json', (req, res) => {
  res.json({ message: 'Hello World!' });
});

app.get('/redirect', (req, res) => {
  res.redirect(301, '/');
});

app.post('/set-cookies', (req, res) => {
  for (const [k, v] of Object.entries(req.body)) {
    res.cookie(k, v);
  }
  res.status(204).send();
});

app.post('/set-cookies-and-redirect-to-get-cookies', (req, res) => {
  for (const [k, v] of Object.entries(req.body)) {
    res.cookie(k, v);
  }
  res.redirect(301, '/get-cookies');
});

app.get('/get-cookies', (req, res) => {
  res.json(req.cookies);
});

app.get('/get-query', (req, res) => {
  res.json(req.query);
});

export async function startServer(port: number): Promise<void> {
  return new Promise(resolve => {
    if (server) {
      resolve();
      return;
    } else {
      server = app.listen(port, resolve);
    }
  });
}

export async function stopServer(): Promise<void> {
  if (!server) return;
  server.close();
  server = undefined;
}


test/index.test.ts
import Fetch from '../src';
import { startServer, stopServer } from './server';

const port = 3000;
const baseUrl = `http://localhost:${port}`;

beforeAll(async () => {
  try {
    await startServer(port);
  } catch (error) {
    process.exit(1);
  }
});

afterAll(async () => {
  await stopServer();
});

describe('advanced-fetch', () => {
  test('can get a simple string', async () => {
    const fetch = new Fetch();
    const response = await fetch.get(`${baseUrl}/`);
    expect(response).toBe('Hello World!');
  });

  test('can get JSON and automatically parses it', async () => {
    const fetch = new Fetch();
    const response = await fetch.get(`${baseUrl}/json`);
    expect(response.message).toEqual('Hello World!');
  });

  test("doesn't parse JSON when told not to", async () => {
    const fetch = new Fetch();
    const response = await fetch.get(`${baseUrl}/json`, { returnType: 'string' });
    expect(response).toEqual('{"message":"Hello World!"}');
  });

  test('returns the URL chain when using manual redirects', async () => {
    const fetch = new Fetch();
    const response = await fetch.requestWithFullResponse(`${baseUrl}/redirect`);

    expect(response.urls).toHaveLength(2);
    expect(response.urls![0]).toBe(`${baseUrl}/redirect`);
    expect(response.urls![1]).toBe(`${baseUrl}/`);
  });

  test('can retain a cookie', async () => {
    const fetch = new Fetch();
    await fetch.post(`${baseUrl}/set-cookies`, { json: { mycookie: 'secret' } });
    const response = await fetch.get(`${baseUrl}/get-cookies`);
    expect(response.mycookie).toBe('secret');
  });

  test('can retain a cookie across a redirect', async () => {
    const fetch = new Fetch();
    const response = await fetch.post(`${baseUrl}/set-cookies-and-redirect-to-get-cookies`, {
      json: { mycookie: 'secret' },
    });
    expect(response.mycookie).toBe('secret');
  });

  test('correctly handles a query parameter', async () => {
    const fetch = new Fetch();
    const response = await fetch.get(`${baseUrl}/get-query`, { query: { hello: 'world' } });
    expect(response.hello).toBe('world');
  });

  test('correctly handles a query parameter appended to one already in the URL', async () => {
    const fetch = new Fetch();
    const response = await fetch.get(`${baseUrl}/get-query?hello=world`, { query: { foo: 'bar' } });
    expect(response.hello).toBe('world');
    expect(response.foo).toBe('bar');
  });

  // TODO: add many more unit tests for cookies, encoding, etc.
});


index.js
const util = require('util');
const fs = require('fs-extra');
const nodeFetch = require('node-fetch');
const { CookieJar } = require('tough-cookie');
const fetchCookie = require('fetch-cookie');
const FormData = require('form-data');
const { URLSearchParams } = require('url');
const querystring = require('querystring');
const iconv = require('iconv-lite');
const https = require('https');

CookieJar.deserialize = util.promisify(CookieJar.deserialize);

function promisifyCookieJar(jar) {
  jar.serialize = util.promisify(jar.serialize);
  jar.getCookies = util.promisify(jar.getCookies);
  return jar;
}

class Fetch {
  constructor({ cookiesFilename, encoding, commonFetchParams, ignoreInvalidHttps } = {}) {
    if (ignoreInvalidHttps) {
      commonFetchParams = {
        ...commonFetchParams,
        agent: new https.Agent({ rejectUnauthorized: false }),
      };
    }

    this.cookiesFilename = cookiesFilename;
    this.encoding = encoding;
    this.commonFetchParams = commonFetchParams;

    return (async () => {
      if (this.cookiesFilename && (await fs.exists(this.cookiesFilename)))
        this.jar = promisifyCookieJar(await CookieJar.deserialize(await fs.readJSON(this.cookiesFilename)));
      else this.jar = promisifyCookieJar(new CookieJar());
      this.fetch = fetchCookie(nodeFetch, this.jar);
      return this;
    })();
  }

  async requestWithHeaders(url, params = {}) {
    // process GET query string
    let queryStr = '';
    if (params.query) {
      if (Object.keys(params.query).length > 0) queryStr = '&' + querystring.stringify(params.query);
      delete params.query;
    }

    // make sure headers object exists
    if (!params.headers) params.headers = {};

    // process request body (params should only have one of "form", "formData" or "json")
    if (params.form) {
      params.body = new URLSearchParams(params.form);
      delete params.form;
    } else if (params.formData) {
      const form = new FormData();
      const appendToForm = (k, v) => {
        if (Array.isArray(v)) v.forEach(x => form.append(k, x.toString()));
        if (typeof v === 'object') form.append(k, v.value, v.options);
        else form.append(k, v.toString());
      };

      for (const [k, v] of Object.entries(params.formData)) {
        if (Array.isArray(v)) v.forEach(vItem => appendToForm(k, vItem));
        else appendToForm(k, v);
      }
      params.body = form;
      delete params.formData;
    } else if (params.json) {
      params.body = JSON.stringify(params.json);
      params.headers['Content-Type'] = 'application/json';
      delete params.json;
    }

    // execute request and store cookies if cookieFilename was given in constructor
    const resp = await this.fetch(url + queryStr, {
      ...params,
      ...this.commonFetchParams,
    });
    if (resp.status >= 500) throw new Error(resp.status + ' ' + resp.statusText);
    await this.storeCookies();

    // if an encoding was given in the Fetch constructur, the response still needs to be converted from that to UTF-8
    let encoding = params.encoding || this.encoding;
    let content = await (params.returnBuffer || encoding ? resp.buffer() : resp.text());
    if (!params.returnBuffer && encoding) content = iconv.decode(content, encoding);

    return {
      status: resp.status,
      headers: resp.headers.raw(),
      content,
    };
  }

  async requestWithFullResponse(url, params = {}) {
    // only let node-fetch follow redirects if that's explicitly stated
    if (params.redirect === 'follow') return (await this.requestWithHeaders(url, params)).content;

    // otherwise follow them manually, because otherwise Set-Cookie is ignored for redirecting sites
    let nextUrl = url;
    let currResp;
    params = { ...params, redirect: 'manual' };
    while (true) {
      currResp = await this.requestWithHeaders(nextUrl, params);
      if (currResp.status < 300 || currResp.status >= 400) break;
      else if (currResp.status !== 301 && currResp.status !== 302)
        throw new Error('unknown HTTP redirect status: ' + currResp.status);

      const loc = currResp.headers.location;
      if (!loc || (Array.isArray(loc) && loc.length !== 1)) break;
      nextUrl = Array.isArray(loc) ? loc[0] : loc;
      params = { method: 'GET' };
    }

    return currResp;
  }

  async request(url, params = {}) {
    return (await this.requestWithFullResponse(url, params)).content;
  }

  get(url, params = {}) {
    return this.request(url, { ...params, method: 'GET' });
  }

  post(url, params = {}) {
    return this.request(url, { ...params, method: 'POST' });
  }

  put(url, params = {}) {
    return this.request(url, { ...params, method: 'PUT' });
  }

  delete(url, params = {}) {
    return this.request(url, { ...params, method: 'DELETE' });
  }

  async storeCookies() {
    if (!this.cookiesFilename) return;
    await fs.writeJSON(this.cookiesFilename, await this.jar.serialize(), {
      spaces: 4,
    });
  }

  async getCookie(key) {
    const cookie = (await this.jar.serialize()).cookies.find(c => c.key === key);
    if (!cookie) return;
    return cookie.value;
  }
}

module.exports = { Fetch };


src/index.ts
import util from 'util';
import fs from 'fs-extra';
import nodeFetch from 'node-fetch-commonjs';
import { CookieJar } from 'tough-cookie';
import fetchCookie from 'fetch-cookie';
import FormData from 'form-data';
import { URLSearchParams } from 'url';
import iconv from 'iconv-lite';
import https from 'https';

CookieJar.deserialize = util.promisify(CookieJar.deserialize);

function promisifyCookieJar(jar: CookieJar): CookieJar {
  jar.serialize = util.promisify(jar.serialize);
  jar.getCookies = util.promisify(jar.getCookies);
  return jar;
}

export interface FetchOptions {
  // cookies can be stored in this file using the `storeCookies` function in the format of `tough-cookie`
  // also, when this filename is given, cookies will be loaded from this file in the first request
  cookiesFilename?: string;

  // global encoding to be used for all requests (can still be overridden in `RequestParams`)
  encoding?: string;

  // options that should always be passed to `node-fetch` calls
  commonFetchParams?: any;

  // ignores issues with HTTPS by setting `rejectUnauthorized: false`
  ignoreInvalidHttps?: boolean;
}

export type FormDataValue = string | string[] | { value: any; options: FormData.AppendOptions | string };

export type RequestReturnType = 'string' | 'json' | 'buffer';

export interface AdvancedFetchRequestParams {
  // HTTP headers
  headers?: { [key: string]: string };

  // GET query parameters, serialized using URLSearchParams
  query?: { [key: string]: string };

  // form data for the body of a `Content-Type: application/x-www-form-urlencoded` request
  form?: { [key: string]: string | string[] };

  // form data for the body of a `Content-Type: multipart/form-data` request
  formData?: { [key: string]: FormDataValue };

  // arbitrary JSON for the body of a `Content-Type: application/json` request
  json?: any;

  // encoding from which the response should be converted to UTF-8
  encoding?: string;

  // whether a string (which is the default), parsed JSON (which is the actual default when
  // the response's Content-Type is application/json) or a Buffer should be returned
  returnType?: RequestReturnType;

  // the default is "followWithCookies", because only then Set-Cookie is respected for redirecting sites
  // otherwise the desired regular node-fetch redirect behavior is used
  redirect?: RequestRedirect | 'followWithCookies';

  // HTTP method
  method?: string;
}

export interface AdvancedFetchResponse {
  // list of URLs that were followed to get to this response (only set when params.redirect was "manual") with the one of the final response last
  // otherwise, this contains either one URL (if no redirects were followed) or two URLs (if a redirect was followed)
  urls: string[];

  // when following redirects, the following properties only refer to the last, final response
  status: number;
  headers: { [key: string]: string };
  content: any;
}

export default class Fetch {
  private options: FetchOptions;
  private jar?: CookieJar;
  private fetch?: any;
  private initialized: boolean = false;

  constructor(options: FetchOptions = {}) {
    this.options = options;

    if (options.ignoreInvalidHttps) {
      options.commonFetchParams = {
        ...options.commonFetchParams,
        agent: new https.Agent({ rejectUnauthorized: false }),
      };
    }
  }

  private async initialize(): Promise<void> {
    if (this.initialized) return;
    if (this.options.cookiesFilename && fs.existsSync(this.options.cookiesFilename))
      this.jar = promisifyCookieJar(await CookieJar.deserialize(await fs.readJSON(this.options.cookiesFilename)));
    else this.jar = promisifyCookieJar(new CookieJar());
    this.fetch = fetchCookie(nodeFetch, this.jar);
    this.initialized = true;
  }

  async requestWithHeaders(url: string, params: AdvancedFetchRequestParams = {}): Promise<AdvancedFetchResponse> {
    await this.initialize();

    // process GET query string
    let queryStr = '';
    if (params.query) {
      if (Object.keys(params.query).length > 0)
        queryStr = (url.includes('?') ? '&' : '?') + new URLSearchParams(params.query).toString();
      delete params.query;
    }

    // make sure headers object exists
    if (!params.headers) params.headers = {};

    // process request body (params should only have one of "form", "formData" or "json")
    let body: any = undefined;
    if (params.form) {
      body = new URLSearchParams(params.form);
      delete params.form;
    } else if (params.formData) {
      const form = new FormData();
      const appendToForm = (k: string, v: FormDataValue) => {
        if (Array.isArray(v)) v.forEach(x => form.append(k, x.toString()));
        else if (typeof v === 'object') form.append(k, v.value, v.options);
        else form.append(k, v.toString());
      };

      for (const [k, v] of Object.entries(params.formData)) {
        if (Array.isArray(v)) v.forEach(vItem => appendToForm(k, vItem));
        else appendToForm(k, v);
      }
      body = form;
      delete params.formData;
    } else if (params.json) {
      body = JSON.stringify(params.json);
      params.headers['Content-Type'] = 'application/json';
      delete params.json;
    }

    // execute request and store cookies if cookieFilename was given in constructor
    const resp = await this.fetch(url + queryStr, {
      body,
      ...params,
      ...this.options.commonFetchParams,
    });
    if (resp.status >= 500) throw new Error(resp.status + ' ' + resp.statusText);
    await this.storeCookies();

    // if an encoding was given in the Fetch constructur, the response still needs to be converted from that to UTF-8
    let encoding = params.encoding || this.options.encoding;
    let content = await (params.returnType === 'buffer' || encoding ? resp.buffer() : resp.text());
    if (params.returnType !== 'buffer' && encoding) content = iconv.decode(content, encoding);
    if (
      params.returnType === 'json' ||
      (resp.headers.get('content-type')?.startsWith('application/json') &&
        params.returnType !== 'buffer' &&
        params.returnType !== 'string')
    ) {
      content = JSON.parse(content);
    }

    return {
      urls: url === resp.url ? [url] : [url, resp.url],
      status: resp.status,
      headers: resp.headers.raw(),
      content,
    };
  }

  async requestWithFullResponse(url: string, params: AdvancedFetchRequestParams = {}): Promise<AdvancedFetchResponse> {
    if (params.redirect === 'followWithCookies') {
      // otherwise follow them manually, in order to also save the cookies set via Set-Cookie by intermediate redirects in the cookie jar
      let nextUrl = url;
      let currOrigin = new URL(url).origin;
      let currResp: AdvancedFetchResponse;
      const manuallyFollowedUrls = [url];
      params = { ...params, redirect: 'manual' };
      while (true) {
        currResp = await this.requestWithHeaders(nextUrl, params);
        if (currResp.status < 300 || currResp.status >= 400) break;
        else if (currResp.status !== 301 && currResp.status !== 302)
          throw new Error('unknown HTTP redirect status: ' + currResp.status);

        const loc = currResp.headers.location;
        if (!loc || (Array.isArray(loc) && loc.length !== 1)) break;
        nextUrl = Array.isArray(loc) ? loc[0] : loc;

        // make sure nextUrl is absolute (if it is, set is as the new origin, otherwise use the same origin like before)
        if (nextUrl.startsWith('/')) nextUrl = currOrigin + nextUrl;
        else currOrigin = new URL(nextUrl).origin;

        manuallyFollowedUrls.push(nextUrl);
        params = { method: 'GET' };
      }

      if (currResp.urls[currResp.urls.length - 1] !== manuallyFollowedUrls[manuallyFollowedUrls.length - 1])
        manuallyFollowedUrls.push(currResp.urls[currResp.urls.length - 1]);

      return { ...currResp, urls: manuallyFollowedUrls };
    }

    return await this.requestWithHeaders(url, params);
  }

  async request(url: string, params: AdvancedFetchRequestParams = {}): Promise<any> {
    return (await this.requestWithFullResponse(url, params)).content;
  }

  get(url: string, params: AdvancedFetchRequestParams = {}): Promise<any> {
    return this.request(url, { ...params, method: 'GET' });
  }

  post(url: string, params: AdvancedFetchRequestParams = {}): Promise<any> {
    return this.request(url, { ...params, method: 'POST' });
  }

  put(url: string, params: AdvancedFetchRequestParams = {}): Promise<any> {
    return this.request(url, { ...params, method: 'PUT' });
  }

  delete(url: string, params: AdvancedFetchRequestParams = {}): Promise<any> {
    return this.request(url, { ...params, method: 'DELETE' });
  }

  async storeCookies(): Promise<void> {
    await this.initialize();

    if (!this.options.cookiesFilename || !this.jar) return;
    await fs.writeJSON(this.options.cookiesFilename, await this.jar.serialize(), {
      spaces: 4,
    });
  }

  async getCookie(key: string): Promise<string | undefined> {
    await this.initialize();

    if (!this.jar) return;
    const cookie = (await this.jar.serialize()).cookies.find(c => c.key === key);
    if (!cookie) return;
    return cookie.value;
  }
}
